### PR TITLE
packagekit: allow dbus activation from snapped apps

### PR DESCRIPTION
**Pull request by sergio-costas**
Currently, a snapped application with the "packagekit-control" plug connected (like snap-store/ubuntu-software), can access the packagekit daemon through DBus without problem... as long as the daemon is already running. If it isn't running, DBus activation won't work for a snapped application.

A way of checking this is by launching a shell inside the snap- store snap

    snap run --shell snap-store

and do a DBus call, like

    dbus-send --system --print-reply --dest=org.freedesktop.PackageKit /org/freedesktop/PackageKit org.freedesktop.DBus.Properties.GetAll string:"org.freedesktop.PackageKit"

If packagekitd is running, this call will succeed, but if it's not running, it will fail.

This problem was detected by the people of the new snap-store. Currently they use a trick to ensure that packagekitd is launched, that consists on:

    dbus-send --system --print-reply --dest=org.freedesktop.DBus /org/freedesktop/DBus org.freedesktop.DBus.StartServiceByName string:org.freedesktop.PackageKit uint32:0

This PR fixes this, although it needs security review.

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

**Review by alexmurray — APPROVED** [review]
LGTM - I wonder if we may have similar problems in other interfaces - but then I guess this PR can serve as an example for how to fix those too if needed.

**Comment by sergio-costas** [comment]
@alexmurray Thanks Alex. Can you put the "needs security review" label, please? I don't have enough privileges to do it.

**Comment by mvo5** [comment]
Thanks! I added and then removed the "needs security" label as Alex already gave his +1 :-)

**Comment by alexmurray** [comment]
> Thanks! I added and then removed the "needs security" label as Alex already gave his +1 :-)

Do we need a 'Security Reviewed' or similar tag which we can add after removing 'Needs Security Review'?

File: interfaces/builtin/packagekit_control.go
Comment (jhenstridge) [inline interfaces/builtin/packagekit_control.go:70]: As an alternative to removing the `peer=` attribute here, one another option would be to create a copy of the rule that uses `peer=(name=org.freedesktop.PackageKit)`.

I don't think we can combine these into a single rule to handle both the "service activation" case and "calling method on the unique name" case.

**Review by jhenstridge — CHANGES_REQUESTED** [review]
Is this a reasonable description of the problem?

1. the gnome-software based store accesses PackageKit via glib proxies that manually activate PackageKit via `StartServiceByName` then communicate with it via its unique name. This all works with the current rules because it never sends method calls to PackageKit until after it has been activated.
2. The new Flutter based store relies on implicit service activation. As there is no `AssumedAppArmorLabel=unconfined` line in PackageKit's service activation file, none of the rules using `peer=(label=unconfined)` will match when the service is not running.

We could allow the traffic for (2) via rules that use `peer=(name=org.freedesktop.PackageKit)`, but they wouldn't match the method calls directed to the unique name as used by (1).

So the right fix is probably to (a) determine what the first method call to PackageKit the Flutter store makes, and (b) add a rule to allow that method call using `peer=(name=org.freedesktop.PackageKit)`. Judging by the changes proposed here, it's either an Introspectable or Properties call. Is that deterministic?

**Comment by sergio-costas** [comment]
@jhenstridge Mmm... Or... if there is a StartServiceByName() call available in flutter, then they should use it and don't touch snapd... I'll talk to them.

**Comment by sergio-costas** [comment]
Ok, they were already doing it that way as a temporary fix, but since it seems to be "the way to go"(tm), then this PR isn't needed. I'll close it.

**Comment by jhenstridge** [comment]
@sergio-costas: for what it is worth, I wasn't saying there was anything wrong with what the Flutter store was doing: just trying to explain why one implementation was fine with the interface as is, and ther other not.

It's entirely valid to rely on implicit service activation, and we can support it in the interfaces. A search for `peer=(name=` will show a number of interfaces already using similar rules.

**Comment by sergio-costas** [comment]
@jhenstridge Fully agree. My point is only that this is a very specific interface, and there is already a valid solution also used by the previous software, so it's better to avoid touching snapd. For a more commonly used interface I would agree in apply a patch to fully support implicit service activation, but for this case I think that it doesn't worth it.

**Pull request closed without merging**
